646 dependency lfric

hiker opened the pull request:
This pull request:

- moves ArgOrdering and derived classes into separate files in domain/lfric
- moved tests for ArgOrdering and derives classes into one test file.
- adds variable access information to these classes, which means that PSyData based transform that rely on dependency analysis (e.g. kernel extraction) will now use all implicit arguments.
- Moves common code in the classes into the base class (#767).

I do have some questions:
1) For now I have moved all ArgOrdering tests into tests/domain/lfric/arg_ordering_test.py. This is clearly not ideal (since it also contains KernCallArgList etc tests), but if I split the tests into individual files, we end up with test files that have (say) only 2 or 3 tests in them. And still none of them would even get close to cover the whole code. Additionally, it actually removed one or two stub generation tests from the stub test file (the ones that test KernStubArgList).
2) As discussed in #767 I have refactored some acc related code to use the base class. This changes the argument order. I admit that I don't fully understand the openacc code yet, so I am wondering if there isn't something else that needs to be changed if the argument order is changed??

hiker commented:
Taking this back for now, since I just learned that (at least) two implicit variables are NOT readonly.

hiker commented:
OK, gravity-wave miniapps now works completely, so ready for review.

hiker commented:
I don't really understand the coverage failure - coverage reduced by 0.07%?? Except ... because of the number of lines removed in dynamo0p3, the still uncovered lines are now a larger fraction? Not sure, am marking it back to ready-for-review.

arporter commented:
Hi @hiker - Iva's PR is now on master so please could you fix the conflicts and then I'll take a look.

hiker commented:
OK, merge done. I had to copy quite a few changes manually, so I hope I didn't miss anything.

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        :param function_space: the function space for which the related \
+            arguments common to LMA operators and fields are added.
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
`psyclone.domain.lfric.FunctionSpace`

hiker replied:
Fixed, in all instance (and I also found a few in dynamo0p3 that I've missed, I've fixed them all).

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        :param function_space: the function space for which the compulsory \
+            arguments are added.
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
domain.lfric.FunctionSpace

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        :type var_accesses: \
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        '''
```
Please add a :raises: or (better?) @abc.abstractmethod.

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        If supplied it also stores this access in var_accesses.
+
+        :param function_space: the function space for which to add arguments
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
psyclone.domain.lfric.FunctionSpace

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        :param function_space: the function space for which the basis \
+                               function is required.
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
domain.lfric

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        :param function_space: the function space for which the differential \
+            basis functions are required.
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
domain.lfric

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        :param function_space: the function space for which orientation \
+            information is added.
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
domain.lfric

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        also stores this access in var_accesses.
+
+        :param function_space: the function space of the operator.
+        :type function_space: :py:class:`psyclone.dynamo3.FunctionSpace`
```
domain.lfric

arporter commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        :param function_space: the function space for which the indirect \
+            dofmap is required.
+        :type function_space: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
domain.lfric.

hiker commented:
Dang, I am completely confused now. This pull request should **NOT** have included any read-only library - neither Fortran files, nor the corresponding PSyclone transforms (since I already split this big chunk of work into two separate pull requests, see #812 ). A big apology from me!

This should only have moved four classes from dynamo0p3 into separate files, added the missing documentation and then the var_access parameter (admittedly the latter could have been a separate PR). It did not change any code except for the ``if var_access:`` parts, which indeed are all new.

I'll try to figure out what happened - maybe we can have a quick chat/call tonight to see how to progress from this?

hiker commented:
I finally figured out what happened: I divided my work correctly into two separate PR, but when it was ready, I merged in the wrong direction - instead of merging this (dependency) into read-only (which needs proper dependency for lfric), I merged read-only into dependency, re-adding all the read-only related code that I had painfully pulled out. Luckily it wasn't too bad, I created a new branch just before the incorrect merge, and re-applied the merges.

I have submitted the new branch as #825. This branch has the changes requested here related to the variable usage detection applied. The comments about the Fortran files and documentation are all related to read-only (#812), and will be applied there.

I'll leave this PR open for now, assuming that it might be easier to see my comments now and later here.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+    def append(self, var_name):
+        '''Appends the specified variable name to the list of all arguments.
+        :param str var_name: the name of the variable.
+E
```
Oops, done.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+    # pylint: disable=too-many-public-methods
+    '''Base class capturing the arguments, type and ordering of data in
+    a Kernel call.
+
```
Done.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+from psyclone.errors import GenerationError, InternalError
+
+
+class ArgOrdering(object):
```
I actually looked at that. The reason why I decided against is that some functions only make sense to call after `generate()` is called (e.g. `[]`, `len()`), but others must be allowed to be called before (`append`, extend()`). Considering that this would then require a fair amount of functions to be overloaded (and I somehow suspect that there are a few), I thought it wasn't worth the effort.

Am happy to redesign this, tbh I liked the idea, till I noticed the `generate()` issue.

arporter replied:
OK - I'm happy if you've looked into it and decided it's not worth it.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+    def arglist(self):
+        '''
+        :return: the kernel argument list. The generate method must be \
+        called first.
```
Done.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        '''
+        if not self._generate_called:
+            raise InternalError(
+                "Internal error. The argument list in KernStubArgList:"
```
Indeed -  a bit more complicated though: `type(self).__name__`, which works in python 2 and 3.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        that can be specialised by a child class for its particular need.
+        If the optional argument var_accesses is supplied, this function
+        will also add variable access information for each implicit argument
+        that is added. These accesses will be marked as read.
```
Done.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+            :py:class:`psyclone.core.access_info.VariablesAccessInfo`
+
+        :raises GenerationError: if the kernel arguments break the
+                                 rules for the Dynamo 0.3 API.
```
Oops.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+    def cell_position(self, var_accesses=None):
+        '''Add cell position information
+
```
Done.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        :raises NotImplementedError: because this is an abstract method
+
+        '''
+        raise NotImplementedError(
```
I left it the way it was originally (and I did notice that one or two functions were declared abstract). But since you suggested it, I am more than happy to make them all proper abstract functions (which then requires to remove one test that checked all these raise statements).
Done.

arporter replied:
Great. If it results in removing a test then I think we're going in the right direction.

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+        :raises NotImplementedError: because this is an abstract method
+
+        '''
+        raise NotImplementedError(
```
Done (same for all other mentions of this issue).

hiker commented on src/psyclone/domain/lfric/arg_ordering.py:
```diff
@@ -0,0 +1,714 @@
+
+        self.append(scalar_arg.name)
+        if var_accesses is not None:
+            var_accesses.add_access(scalar_arg.name, AccessType.READ,
```
Oh yes indeed, big bug that. I've fixed it, but `grep` did not find any test case that uses a written scalar value. Do we have any test for this??

hiker commented:
I've just merged #825 as ready, which also includes the fixes related to psyclone/domain/lfric files (and tests), but NOT the fixed to anything related to read-only (which I will do later in #812). I'll leave this PR as 'in progress', till #825 and #812 are merged.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+    --------------------------------------
+
+The transformation that adds read-only-verification to an application
+can be applied for any API. Here an example that searches for each
```
Done.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+can be applied for any API. Here an example that searches for each
+loop in an invoke (which will always surround kernel calls) and applies the
+transformation to each one. This code has been successfully used as a
+global transformation with the LFRic gravity-wave miniapps::
```
Done.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+Both libraries support the environment variable ``PSYDATA_VERBOSE``
+at runtime, which can be used to control how much output is generated
+by the read-only-verification library at runtime. If the
+variable is not specified or has the value '0', only changed checksums
```
Done.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+
+.. note:
+    It is the responsibility of the user to make sure that the infrastructure
+    files used during compilation of the Read-Only-Verification library is
```
Done.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+Read-Only-Verification Library for GOcean
++++++++++++++++++++++++++++++++++++++++++
+This library is contained in ``lib/read_only/dl_esm_info`` and it
+must be compiled before linking any LFRic-based application that uses
```
Oops

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+``r2d_field`` class).
+
+The makefile uses the variable ``DL_DIR`` to point to the location where
+LFRic's ``field_mod`` has been compiled. It defaults to
```
Oops again.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+The makefile uses the variable ``DL_DIR`` to point to the location where
+LFRic's ``field_mod`` has been compiled. It defaults to
+``../../../external/dl_esm_inf/finite_difference/src``,
+which is the location of the dl_esm_info version that is included with
```
Done.

hiker commented on doc/user_guide/psy_data.rst:
```diff
@@ -80,3 +82,123 @@ the PSyData API, for example ``ProfileTrans``, ``GOceanExtractTrans``
+
+.. note:
+    It is the responsibility of the user to make sure that the infrastructure
+    files used during compilation of the Read-Only-Verification library is
```
Done.

hiker commented on examples/gocean/eg7/Makefile:
```diff
@@ -0,0 +1,74 @@
+
+GENERATED_FILES += alg.f90 psy.f90 *.o *.mod
+
+DL_DIR = ../../../external/dl_esm_inf/finite_difference/src
```
Indeed. I've also applied some other changes from eg5/6 - e.g. DL_DIR is now INF_DIR etc.

hiker commented on examples/gocean/eg7/Makefile:
```diff
@@ -0,0 +1,74 @@
+	$(MAKE) -C $(READ_ONLY_DIR)
+
+alg.f90 psy.f90: test.x90 read_only_transform.py
+	psyclone -nodm -l -api "gocean1.0" -s ./read_only_transform.py\
```
Great!

hiker commented on examples/gocean/eg7/Makefile:
```diff
@@ -0,0 +1,74 @@
+
+transform: alg.f90
+
+compile: $(READ_ONLY_DIR)/$(LIB_NAME) $(KERNELS) alg.o psy.o
```
Yes indeed (actually, already the kernels need to depend on $(INF_LIB), since the .mod files are required for compilation).

hiker commented on examples/gocean/eg7/read_only_transform.py:
```diff
@@ -0,0 +1,74 @@
+
+def trans(psy):
+    '''
+    Take the supplied psy object, and add kernel extraction code.
```
Done.

hiker commented on examples/gocean/eg7/test.x90:
```diff
@@ -0,0 +1,40 @@
+
+    TYPE(r2d_field)          :: a_fld, b_fld, c_fld, d_fld
+    double precision         :: x, y
+    ! Just to show that different declarations work as expected
```
Oops.

hiker commented on examples/gocean/eg7/test.x90:
```diff
@@ -0,0 +1,40 @@
+                 init_field(c_Fld, 3.0), &
+                 init_field(d_Fld, 4.0)    )
+
+    x = 0
```
I was just wondering why it compiled, since I had problems with that in the past, but I remember that that was then calling a generic function - so I just learned that in generic calls no automatic cast from int to double happens, while it does in an assignment ;) Anyway, 0.0d0 is of course clearer, so I've changes this :)

hiker commented on examples/gocean/eg7/test.x90:
```diff
@@ -0,0 +1,40 @@
+    real(kind=kind(1.0d0))   :: z
+    TYPE(grid_type), target  :: grid
+
+    call parallel_init()
```
Done, added comments across the whole program.

hiker commented on examples/gocean/eg7/test.x90:
```diff
@@ -0,0 +1,40 @@
+    z = 1
+    call invoke (update_field(a_fld, b_fld, c_fld, d_fld, x, y, z) )
+
+    print *,b_fld%data(1, 1)
```
I have removed this print statement. The reason I usually add a print statement is that sometimes the compiler gets very clever and removes the whole computation if the output is not used. But in this case we have output via the psydata library, so the compiler cannot remove the loops :)

hiker commented on examples/gocean/eg7/update_field_mod.f90:
```diff
@@ -0,0 +1,95 @@
+             go_arg(GO_READ,      GO_CT,       GO_POINTWISE),  & ! field
+             go_arg(GO_READ,      GO_CT,       GO_POINTWISE),  & ! field
+             go_arg(GO_READ,      GO_CT,       GO_POINTWISE),  & ! field
+             go_arg(GO_READWRITE, GO_R_SCALAR, GO_POINTWISE),  & ! scalar
```
I'll leave this as a note for you, and didn't do anything :) From memory I added the three various access modes to test the variable access detection. I'll see what I need to do once you have added tests that will prohibit this from happening.

hiker commented on lib/read_only/dl_esm_inf/Makefile:
```diff
@@ -0,0 +1,57 @@
+
+F90 ?= gfortran
+
+DL_DIR = ../../../external/dl_esm_inf/finite_difference/src
```
Indeed.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+                                                 DeclareFieldDouble
+
+        !> The generic interface for providing the value of variables,
+        !! which in case of the NetCDF interface is written:                                               
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    !! @param[inout] this The instance of the PSyDataType.
+    !! @param[in] module_name The name of the module of the instrumented
+    !!            region.
+    !! @param[in] kernel_name The name of the instrumented region.
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+            else if (verbose=="2") then
+                this%verbosity = 2
+            else
+                print *,"PSyData: invalid setting of PSYDATA_VERBOSE."
```
I won't comment this in other places: I've changed _all_ print statements to write to stderr in both read-only-libraries.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+        endif
+
+        if (num_pre_vars /= num_post_vars) then
+            print *,"The same number of variables must be provided before and"
```
Done everywhere (in the actual warning messages I am using
```
---------------- PsyData --------------
```
as header and footer.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    end subroutine PostEnd
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a scalar integer value.
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine either computes a checksum or compares a checksum
+    !! (depending on this%verify_checksums)
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    end subroutine ChecksumScalarInt
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a scalar single precision value.
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+        integer(kind=int64) :: chksum64
+        integer(kind=int32) :: chksum32
+
+        ! Type-case from real to 32-bit int (same size, so
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    end subroutine DeclareScalarReal
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine either computes a checksum or compares a checksum
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    end subroutine ChecksumScalarReal
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a scalar single precision value.
```
Done.

hiker replied:
(though I am wondering if we should add a special flag 'no declaration' to prevent the creation of the declaration calls. That would be similar to what happens implicitly already in profile - no declare or provide calls are done if no variable is printed). Well, enhancement for later.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+        double precision, intent(in) :: value
+    end subroutine DeclareScalarDouble
+    ! -------------------------------------------------------------------------
+    !> This subroutine either computes a checksum or compares a checksum
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+                ! Convert 64 bit integer back to 64 bit double precision:
+                orig_value = transfer(this%checksums(this%next_var_index), orig_value)
+                print *,"--------------------------------------"
+                print *,"Double precision variable ", name, " has been modified in ", &
```
I agree - all are going to stderr now.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a double precision field as defined in
+    !! dl_esm_info (r2d_field). A corresponding variable definition is added
```
Oops.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    end subroutine DeclareFieldDouble
+
+    ! -------------------------------------------------------------------------
+    function ComputeChecksum(field) result(checksum)
```
Now that was embarrassing :(

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+    end function ComputeChecksum
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine writes the value of a dl_esm_field (r2d_field)
```
Done.

hiker commented on lib/read_only/dl_esm_inf/read_only.f90:
```diff
@@ -0,0 +1,416 @@
+        cksum = ComputeChecksum(value%data)
+        if(this%verify_checksums) then
+            if(this%checksums(this%next_var_index) /= cksum) then
+                print *,"--------------------------------------"
```
Indeed.

hiker commented on lib/read_only/lfric/Makefile:
```diff
@@ -0,0 +1,54 @@
+# export F90FLAGS="-O3"
+# export AR=ar
+
+LFRIC_DIR ?= ../../../../../lfric/work/trunk/miniapps/gravity_wave/working/field
```
Done.

hiker commented on lib/read_only/lfric/Makefile:
```diff
@@ -0,0 +1,54 @@
+
+$(LIB_NAME):	read_only.o
+	${AR} ${ARFLAGS} ${LIB_NAME} $^
+
```
I tried - I am not sure about doing this case-insensitive, so I only hard-coded the mod name in lowercase.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+! -----------------------------------------------------------------------------
+! Authors J. Henrichs, Bureau of Meteorology
+
+!> This module implements a verification that read-only fields are
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+
+module read_only_verify_psy_data_mod
+    use, intrinsic :: iso_fortran_env, only : int64, int32
+        use field_mod, only : field_type
```
I must have been on some happy juice when writing this line.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+                                                 DeclareFieldVectorDouble
+
+        !> The generic interface for providing the value of variables,
+        !! which in case of the NetCDF interface is written:                                               
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    !!            region.
+    !! @param[in] kernel_name The name of the instrumented region.
+    !! @param[in] num_pre_vars The number of variables that are declared and
+    !!            written before the instrumented region.
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+            else if (verbose=="2") then
+                this%verbosity = 2
+            else
+                print *,"PSyData: invalid setting of PSYDATA_VERBOSE."
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        endif
+
+        if (num_pre_vars /= num_post_vars) then
+            print *,"PSYDATA: The same number of variables must be provided before"
```
I checked, LFRic is using stop at this stage, I couldn't find anything else. There is a log call that will abort, but I am not sure if we want to add this dependency to log(?) - let me know.

For gocean I used the corresponding call.

hiker replied:
(to explain my issues with log: iirc it would be best if the psydata libraries only depend on infrastructure code - and imho I am not sure if log should be part of that).

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        ! stored was counted, so allocate the array now (if it has
+        ! not been allocated already in a previous call):
+        if (.not. allocated(this%checksums)) then
+            allocate(this%checksums(this%count_checksums))
```
All done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+
+        if (this%verify_checksums) then
+            if (value /= this%checksums(this%next_var_index)) then
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        implicit none
+        class(read_only_verify_PSyDataType), intent(inout), target :: this
+        character(*), intent(in) :: name
+        integer, intent(in) :: value
```
I don't agree with this: LFRic has 5 different integer types: 
```
  integer, parameter :: i_def        = int32       !< Default integer kind for application.
  integer, parameter :: i_native     = kind(i_val) !< Native kind for integer.
  integer, parameter :: i_long       = int64       !< Explicit long integer.
  integer, parameter :: i_halo_index = int64       !< Integer kind for the index used in halo swapping
  integer, parameter :: i_um         = int32       !< Default integer kind used by the UM.
```
Since they can all be different, we would have to supply a separate version for each ``i_`` value. On the other hand, if two ``i_`` are the same, compilation will fail because the two functions can not be differentiated.

So I think the right solution is to just provide the standard integers, independent of how LFRic assign them. If they use int*8, we will add them as int*8 (well, whatever the proper portable declaration in modern Fortran is). Note that this will be trivial if we use scripts to create the functions ... see #832 which I just added.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        implicit none
+        class(read_only_verify_PSyDataType), intent(inout), target :: this
+        character(*), intent(in) :: name
+        integer, intent(in) :: value
```
(OK, I won't reply to all the other i_def comments, we can just discuss this above, and I'll adopt it everywhere).

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    end subroutine ChecksumArray2dInt
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares an array of integer value.
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        do k=1, size(value, 3)
+            do j=1, size(value, 2)
+                do i=1, size(value, 1)
+                    checksum = checksum + value(i, j, k)
```
I don't think so, the promotion from 32-bit int to 64-bit happens automatically by the compiler.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        checksum = 0
+        do j=1, size(value, 2)
+            do i=1, size(value, 1)
+                checksum = checksum + value(i, j)
```
As above. Or do you prefer to cast to make this obvious?

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        enddo
+        if (this%verify_checksums) then
+            if (checksum /= this%checksums(this%next_var_index)) then
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    end subroutine DeclareScalarReal
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine either computes a checksum or compares a checksum
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        class(read_only_verify_PSyDataType), intent(inout), target :: this
+        character(*), intent(in) :: name
+        real, intent(in) :: value
+        real :: orig_value
```
Same argument as for ``i_def`` I'd say. In #832 we will add the ability to automatically create single- and double-precision versions, independent of how many types of real LFRic might create in the future (esp. with mixed precision coming).

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        integer(kind=int64) :: chksum64
+        integer(kind=int32) :: chksum32
+
+        ! Type-case from real to 32-bit int (same size, so
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+                ! Convert back to 32 bit, and then cast to real:
+                chksum32 = this%checksums(this%next_var_index)
+                orig_value = transfer(chksum32, orig_value)
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        checksum = ComputeChecksum3d(value)
+        if(this%verify_checksums) then
+            if(this%checksums(this%next_var_index) /= checksum) then
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    end subroutine ChecksumArray4dDouble
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a scalar single precision value.
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        checksum = ComputeChecksum4d(value)
+        if(this%verify_checksums) then
+            if(this%checksums(this%next_var_index) /= checksum) then
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+            if(this%checksums(this%next_var_index) /= cksum) then
+                ! Convert 64 bit integer back to 64 bit double precision:
+                orig_value = transfer(this%checksums(this%next_var_index), orig_value)
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    !! @param[in] value The value of the variable.
+    !! @param[inout] this The instance of the read_only_verify_PSyDataType.
+    subroutine DeclareFieldDouble(this, name, value)
+!        use field_mod, only : field_type
```
Oops.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        cksum = ComputeChecksum1d(value_proxy%data)
+        if(this%verify_checksums) then
+            if(this%checksums(this%next_var_index) /= cksum) then
+                print *,"--------------------------------------"
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    end subroutine ChecksumFieldDouble
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine declares a vector of double precision field based on
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    !! @param[in] value The value of the variable.
+    !! @param[inout] this The instance of the read_only_verify_PSyDataType.
+    subroutine DeclareFieldVectorDouble(this, name, value)
+!        use field_mod, only : field_type
```
Ooops ... again.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+        implicit none
+        class(read_only_verify_PSyDataType), intent(inout), target :: this
+        character(*), intent(in) :: name
+        type(field_type), dimension(3), intent(in) :: value
```
Indeed, I missed that. I think in all other places I have ``:`` already. I had some problems with matching the generic interfaces, and thought that ``:`` vs ``3`` makes a difference ... it didn't, but then I forgot to change it back here.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine computes the checksums for a vector of fields based on
+    !! the LFRic infrastructure type field_type. Depending on state
```
Not really, it does both, but it does rely on the single-field variation. I've updated the comment to make this clearer.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+    end subroutine ChecksumFieldVectorDouble
+
+    ! -------------------------------------------------------------------------
+    !> This function computes a 64-bit integer checksum for a 3d double
```
Done.

hiker commented on lib/read_only/lfric/read_only.f90:
```diff
@@ -0,0 +1,835 @@
+
+    ! -------------------------------------------------------------------------
+    
+subroutine test(my_field)
```
Indeed not.

hiker commented:
OK, I believe I have addressed all comments now, though as stated in the two individual PR for 199 and 646. This PR can be closed once you are looking at #812, which has all changes related to read-only verification applied, and have resolved the comments here. Is there anything I can do to make the switch easier?
